### PR TITLE
[RDY] menu_get() / export menus

### PR DESCRIPTION
**Pull request by teto**
adds export_menu that saves menus to a dict.
`let l = export_menus("")`
With json_encode, this allows UIs to retrieve the menus (likewise for plugins, eg, a certain "palette")

Crashes for now.

### CODE REVIEW

**Comment by teto** [comment]
Removed the crash.
In my init.vim, I have:
`menu Spell.FR :setlocal spell spelllang=fr_fr<CR>`
If you run :echo export_menus(""); it get exported as `"ncommand": ":setlocal spell spelllang=fr_fr\r"`, I wonder if the \r can be a problem ?

**Comment by teto** [comment]
I use a function to test locally:
```
function! ExportMenus(path, modes)
	" export all for now
	let m = export_menus(a:path, 'n')
	let r =  json_encode(m)
	" put =r " to display in current buffer
	call writefile([r], "menus.txt")
endfunc
```
So far it generates:
```
[
    {
        "enabled": 0,
        "hidden": 0,
        "hint": 0,
        "name": "Spell",
        "priority": 500,
        "shortcut": 0,
        "submenus": [
            {
                "enabled": 0,
                "hidden": 0,
                "hint": 0,
                "implementations": [
                    {
                        "noremap": 0,
                        "rhs": ":setlocal spell spelllang=en_us | call histadd('cmd', 'setlocal spell spelllang=en_us')\r",
                        "sid": 0,
                        "silent": 0
                    }
                ],
                "name": "EN_US",
                "priority": 500,
                "shortcut": 0
            },
            {
                "enabled": 0,
                "hidden": 0,
                "hint": 0,
                "implementations": [
                    {
                        "noremap": 0,
                        "rhs": ":setlocal spell spelllang=fr_fr\r",
                        "sid": 0,
                        "silent": 0
                    }
                ],
                "name": "FR",
                "priority": 500,
                "shortcut": 0
            }
        ]
    },
    {
        "enabled": 0,
        "hidden": 0,
        "hint": 0,
        "name": "Trans",
        "priority": 500,
        "shortcut": 0,
        "submenus": [
            {
                "enabled": 0,
                "hidden": 0,
                "hint": 0,
                "implementations": [
                    {
                        "noremap": 0,
                        "rhs": ":te trans :fr <cword>\r",
                        "sid": 0,
                        "silent": 0
                    }
                ],
                "name": "FR",
                "priority": 500,
                "shortcut": 0
            }
        ]
    },
    {
        "enabled": 0,
        "hidden": 0,
        "hint": 0,
        "name": "Search",
        "priority": 500,
        "shortcut": 0,
        "submenus": [
            {
                "enabled": 0,
                "hidden": 0,
                "hint": 0,
                "implementations": [
                    {
                        "noremap": 0,
                        "rhs": ":exe Grepper -grepprg rg --vimgrep $* $.",
                        "sid": 0,
                        "silent": 0
                    }
                ],
                "name": "CurrentBuffer",
                "priority": 500,
                "shortcut": 0
            },
            {
                "enabled": 0,
                "hidden": 0,
                "hint": 0,
                "implementations": [
                    {
                        "noremap": 0,
                        "rhs": ":exe Grepper -grepprg rg --vimgrep $* $+",
                        "sid": 0,
                        "silent": 0
                    }
                ],
                "name": "AllBuffers",
                "priority": 500,
                "shortcut": 0
            }
        ]
    },

```
It doesn't handle icons as neovim removed support for icons and it feels too much to have neovim handle icons IMO.

if that's ok I can proceed to clean up & write the tests ? any advice as how to write the advice ? I believe I can compare a lua table against the expored json ?
```
    local m = funcs.export_menus("","")
    m = funcs.json_encode(m)
```

**Comment by justinmk** [comment]
If gVim has icon support for menus, we should that (at least as far as providing a field in the dictionary and `:menu`; how/if it is supported is decided by UIs). Not necessarily in this PR, but just FYI.

>  I believe I can compare a lua table against the expored json ?

If you get the map through a API function (e.g. `get_menu()`, just like `get_keymap()` in @tjdevries PR), it will automatically marshall to a Lua table which can be compared using `helpers.eq()`.

**Comment by justinmk** [comment]
`export_menus()` doesn't seem the right name. Currently we have `foo_encode()`, `foo_dump()`, and `foo_get()`. Let's not add "export". I would think `menu_get()` is fine, as Bram mentioned in the Vim ticket.

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -5449,6 +5449,30 @@ max({list})	Return the maximum value of all items in {list}.
 		be used as a Number this results in an error.
 		An empty |List| results in zero.
 
+menu_get({path}, {modes})				*dump_menu()*
```
Comment (teto) [inline runtime/doc/eval.txt]: dump_menu => menu_get

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -5449,6 +5449,30 @@ max({list})	Return the maximum value of all items in {list}.
...
+should produce:
+<
+[
+{'hidden': 0, 'submenus':
+  [{'hidden': 0, 'name': 'Test', 'priority': 500, 'shortcut': 0, 'implementations':
+ {'s': {'noremap': 1, 'enabled': 1, 'silent': 0, 'sid': 1, 'rhs': 'x'}, 'v': {'noremap': 1, 'enabled': 1, 's
+ilent': 0, 'sid': 1, 'rhs': 'x'}, 'i': {'noremap': 1, 'enabled': 1, 'silent': 0, 'sid': 1, 'rhs': 'insert'},
+ 'n': {'noremap': 1, 'enabled': 1, 'silent': 0, 'sid': 1, 'rhs': 'inormal'}}}], 'name': 'Test', 'priority':
+500, 'shortcut': 84}
+]
```
Comment (teto) [inline runtime/doc/eval.txt]: ZyX-l suggested to add an example, too bad we can't generate it from the code. is there any vim-prettify that I could use to properly format this part ?

File: src/nvim/eval.lua [hunk 3]
```diff
@@ -85,6 +85,7 @@ return {
     exists={args=1},
     exp={args=1, func="float_op_wrapper", data="&exp"},
     expand={args={1, 3}},
+    menu_get={args={1, 2}},
```
Comment (teto) [inline src/nvim/eval.lua]: Need to put in alphabetical order.

**Review by teto — COMMENTED** [review]
I will complete the tests when the form of the exported dict seems correct, maybe the mnemonic should be converted back to a chracter ?

File: runtime/doc/eval.txt (hunk 2 above)
Comment (justinmk) [inline runtime/doc/eval.txt]: Heh. Sadly `gqq` can't format Vim's own syntax. But I would:

1. `J`oin it to one line
2. `json_encode()` it: `:put =json_encode(<C-R>0)`
3. Format it:
    ```
    :setl formatprg=python\ -m\ json.tool
    gqq
    ```

(If it's just showing a generic map, doesn't matter if it's JSON or VimL, I suppose)

**Comment by ZyX-I** [comment]
AFAIR it was not me who suggested to add an example. Actually documentation I write tends to be more “raw description” then something with examples. (I am not saying that this is how documentation should be written, only how I normally write it.)

As for how you get the results: use `[helpers.]funcs.menu_get(args)` for VimL functions like now. I am rather unsure about adding API function in the PR: calling VimL function is basically just as easy and it makes overhead (converting VimL value while you can just create Array directly, theoretically even create msgpack directly (need serious refactoring, see #4971)) more obvious.

**Comment by teto** [comment]
Maybe I misunderstood; from https://github.com/vim/vim/issues/1563:

> An example in the docs would be very helpful.

whatever, the example is helpful anyway. I don't think adding API functions is helpful either. 

Without further comments, I consider the PR ready (functional tests pass locallly, hopes it passes here), icons & hints can be added later when UIs need it.

**Comment by ZyX-I** [comment]
@teto You are referring to https://github.com/vim/vim/issues/1563#issuecomment-290821247? This is comment from Bram, he is quoting me.

**Comment by teto** [comment]
I made a poc in this plugin (not optimized, still brittle but works locally)  https://github.com/teto/nvim-palette. While waiting for https://github.com/neovim/neovim/pull/6288, I embed a copy  https://github.com/teto/nvim-palette/tree/master/data in the plugin so the palette can load option descriptions. 
Then if you use this PR with the plugin, you should have :PaletteMenu (only menu entries) working and :Palette should regroup both options and menu entries.

File: runtime/doc/eval.txt [hunk 4]
```diff
@@ -5506,6 +5506,66 @@ max({expr})	Return the maximum value of all items in {expr}.
...
+
+    For instance, executing:
+<
+    nnoremenu &Test.Test inormal
+    inoremenu Test.Test insert
+    vnoremenu Test.Test x
+    echo menu_get("")
+>
+should produce an output with a similar structure:
+<
```
Comment (justinmk) [inline runtime/doc/eval.txt]: I think these `> ... <` pairs are reversed.

```
    For instance, executing: >
     nnoremenu &Test.Test inormal
     inoremenu Test.Test insert
     vnoremenu Test.Test x
     echo menu_get("")
<
    should produce an output with a similar structure: >

Comment (justinmk) [inline runtime/doc/eval.txt]: The sample output in both cases could be abbreviated like this:

```
{
    "hidden": 0,
    "name": "Test",
    "priority": 500,
    "shortcut": 84,
    "submenus": [
        {
            "hidden": 0,
            "implementations": {
                "i": {
                    "enabled": 1,
                    "noremap": 1,
                    "rhs": "insert",
                    "sid": 1,
                    "silent": 0
                },
                "n": { ... },
                "s": { ... },
                "v": { ... }
            },
            "name": "Test",
            "priority": 500,
            "shortcut": 0
        }
    ]
}
```

**Comment by justinmk** [comment]
Instead of "implementations", "definitions" or "mappings" would be a more familiar name (if either of those makes sense in context).

**Comment by justinmk** [comment]
I don't see any reason to keep this waiting. Anyone else have comments?

File: runtime/doc/eval.txt (hunk 4 above)
Comment (teto) [inline runtime/doc/eval.txt]: several things were wrong with the doc ('>' pairs, spaces vs tabs, json format etc). I updated it and checked, I think it's fine now.

**Comment by teto** [comment]
rebased, renamed "implementations" to "mappings".

**Comment by justinmk** [comment]
merged. nicely done!

**Pull request closed without merging**